neps/nep-0364.md

```diff
@@ -0,0 +1,188 @@
+
+```
+ed25519: took 9926ms
+```
+
+As an extra data point, when passing `--enable-simd` instead of `--singlepass`
+
+```
+ed25519: took 3085ms
+```
+
+Steps to reproduce:
+commit: `31cf97fb2e155d238308f062c4b92bae716ac19f` in `https://github.com/blasrodri/near-test`
+
+```sh
+# wasi singlepass
+cargo wasi build --bin benches --release
+wasmer compile --singlepass ./target/wasm32-wasi/release/benches.wasi.wasm -o benches_singlepass
+wasmer run ./benches_singlepass
+```
+
+```sh
+# rust native
+cargo run --bin benches --release
+```
+
+Overall: the difference between the two versions (native vs wasi + singlepass is)
+
+```
+ed25519: 25.64x slower
+```
+
+### What is the impact of not doing this?
+
+Costs of running IBC-compatible trustless bridges would be very high. Plus, the fact that signature verification
+is such an expensive operation will force the contract to split the process of batch verification of signatures
+into multiple transactions.
+
+### Why is this design the best in the space of possible designs?
+
+Adding existing proved and vetted crypto crates into the runtime is a safe workaround. It will boost performance
+between 20-25x according to our benchmarks. This will both reduce operating costs significantly and will also
+enable the contract to verify all the signatures in one transaction, which will simplify the contract design.
+
+### What other designs have been considered and what is the rationale for not choosing them?
+
+One possible alternative would be to improve the runtime implementation so that it can compile WASM code to a sufficiently
+fast machine code. Even when it may not be as fast as LLVM native produced code it could still be acceptable for
+these types of use cases (CPU intensive functions) and will certainly avoid the need of adding host functions.
+The effort of adding such a feature will be significantly higher than adding these host functions one by one.
+But on the other side, it will decrease the need of including more host functions in the future.
+
+Another alternative is to deal with the high cost of computing/verifying these signatures in some other manner.
+Decreasing the overall cost of gas and increasing the limits of gas available to attach to the contract could be a possibility.
+Introducing such modification for some contracts, and not for some others can be rather arbitrary
+and not straightforward in the implementation, but an alternative nevertheless.
+
+## Specification
+
+This NEP aims to introduce the following host function:
+
+```rust
+/// Ed25519 is a public-key signature system with several attractive features
+///
+/// Proof of Stake Validator sets can contain different signature schemes.
+/// Ed25519 is one of the most used ones across blockchains, and hence it's importance to be added.
+/// For further reference, visit: https://ed25519.cr.yp.to
```

Comment:
I wonder if we should add more inline information about how `ed25519` works, in particular how it handles some of the edge cases. An alternative to add this information inline would be to add this information to [Nomicon](https://nomicon.io/), given `ed25519` is already a cyptographic primitive used in NEAR before this NEP.

The motivation is to allow potential new client implementations to be able to build the client by only reading Nomicon and NEPs.